uaclient/contract.py

```diff
@@ -246,6 +244,92 @@ def get_contract_token_for_account(contract_client, macaroon, account_id):
+    ent_cls = entitlements.ENTITLEMENT_CLASS_BY_NAME[name]
+    if not issubclass(ent_cls, RepoEntitlement):
+        return
+    if deltas.get('entitled') is False:  # Then refresh disabled support
+        logging.debug(
+            "Due to contract refresh, '%s' is now disabled.", name)
+        entitlement = ent_cls()
+        entitlement.disable()
+        # Clean up former entitled machine-access-<name> response cache file
+        # because uaclient doesn't access machine-access-* routes or responses
+        # on unentitled services.
+        entitlement.cfg.delete_cache_key('machine-access-%s' % name)
+    new_token = deltas.get('resourceToken')
+    new_url = deltas.get('directives', {}).get('aptURL')
+    if new_token or new_url:
+        logging.info(
+            "Updating '%s' apt sources list on changed directives.", name)
+        series = util.get_platform_info('series')
+        repo_filename = ent_cls.repo_list_file_tmpl.format(
+            name=ent_cls.name, series=series)
+        if not new_token:
+            new_token = orig_access['resourceToken']
+        creds = 'bearer:%s' % new_token
+        if new_url:
+            # Remove original aptURL and auth and rewrite
+            apt.remove_auth_apt_repo(
+                repo_filename, orig_access['directives']['aptURL'])
+        else:
+            new_url = orig_access.get('directives', {}).get('aptURL')
+        suites = deltas.get('directives', {}).get('suites', [])
+        if not suites:
+            suites = orig_access.get('directives', {}).get('suites', [])
+        apt.add_auth_apt_repo(repo_filename, new_url, creds, suites=suites)
+
+
+def process_entitlement_delta(orig_access, new_access):
+    """Process a entitlement access dictionary deltas if they exist."""
+    if not orig_access or orig_access == new_access:
+        return
+    deltas = {}
+    name = orig_access['entitlement']['type']
+    orig_token = orig_access.get('resourceToken')
+    if orig_token and orig_token != new_access.get('resourceToken'):
+        logging.debug("Resource token changed for '%s'.", name)
+        deltas['resourceToken'] = new_access.get('resourceToken')
+
+    new_entitlement = new_access['entitlement']
+    if orig_access['entitlement']['entitled'] != new_entitlement['entitled']:
+        deltas['entitled'] = new_entitlement['entitled']
+    new_affordances = new_entitlement.get('affordances', {})
+    orig_affordances = orig_access['entitlement'].get('affordances', {})
+    for affordance, value in orig_affordances.items():
+        if value != new_affordances.get(affordance):
+            logging.debug(
+                "Affordance value for '%s.%s' changed to '%s'",
+                name, affordance, new_affordances.get(affordance))
+            if 'affordances' not in deltas:
+                deltas['affordances'] = {}
+            deltas['affordances'][affordance] = new_affordances.get(affordance)
+    new_directives = new_entitlement.get('directives', {})
```

Comment:
The entitlements code and the directives code are almost identical; could it be refactored in to a single helper that we call twice from this function?